mlir/lib/IR/PatternMatch.cpp

```diff
@@ -317,7 +317,16 @@ void RewriterBase::inlineBlockBefore(Block *source, Block *dest,
 
   // Move operations from the source block to the dest block and erase the
   // source block.
-  dest->getOperations().splice(before, source->getOperations());
+  if (!listener) {
+    // Fast path: If no listener is attached, move all operations at once.
+    dest->getOperations().splice(before, source->getOperations());
+  } else {
+    while (!source->empty())
```

Comment:
OOC what is the overhead here vs doing a splice and then invoking notify inserted? (Probably neglible but just checking)

Reply:
I didn't measure it but I expect the current implementation to be slightly slower than a `splice` followed by a loop that sends all the notifications.

However, the current implementation is the only safe way to implement it. The `notifyOperationInserted` notification has a `Block::iterator` for the previous location of the op, and when moving more than one op at once, it may not be possible to pass a valid `Block::iterator` for the previous location of the first op. (Moving the second op could invalidate the iterator for the first op, or make it point somewhere else.)

I added the fast path to make sure that we have the same performance as before in most cases.